pkg/user/cache/groups.go

```diff
@@ -0,0 +1,81 @@
+package cache
+
+import (
+	"fmt"
+	"time"
+
+	kapi "github.com/GoogleCloudPlatform/kubernetes/pkg/api"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/client/cache"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/fields"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/labels"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/runtime"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/watch"
+
+	userapi "github.com/openshift/origin/pkg/user/api"
+	groupregistry "github.com/openshift/origin/pkg/user/registry/group"
+)
+
+type GroupCache struct {
+	indexer   cache.Indexer
+	reflector *cache.Reflector
+}
+
+const byUserIndexName = "ByUser"
+
+func ByUserIndexKeys(obj interface{}) ([]string, error) {
+	group, ok := obj.(*userapi.Group)
+	if !ok {
+		return []string{""}, fmt.Errorf("unexpected type: %v", obj)
+	}
+
+	return group.Users, nil
+}
+
+func NewGroupCache(groupRegistry groupregistry.Registry) *GroupCache {
+	allNamespaceContext := kapi.WithNamespace(kapi.NewContext(), kapi.NamespaceAll)
+
+	indexer := cache.NewIndexer(cache.MetaNamespaceKeyFunc, cache.Indexers{byUserIndexName: ByUserIndexKeys})
+	reflector := cache.NewReflector(
+		&cache.ListWatch{
+			ListFunc: func() (runtime.Object, error) {
+				return groupRegistry.ListGroups(allNamespaceContext, labels.Everything(), fields.Everything())
+			},
+			WatchFunc: func(resourceVersion string) (watch.Interface, error) {
+				return groupRegistry.WatchGroups(allNamespaceContext, labels.Everything(), fields.Everything(), resourceVersion)
+			},
+		},
+		&userapi.Group{},
+		indexer,
+		2*time.Minute,
```

Comment:
how was this timeout chosen? copy/paste? not sure what we consider best practice for these timeouts now... max watch duration or forced shorter refresh